stdlib/public/BackDeployConcurrency/TaskPool.cpp

```diff
@@ -0,0 +1,872 @@
+
+    PoolStatus statusCancel() {
+      auto old = status.fetch_or(PoolStatus::cancelled,
+                                 std::memory_order_relaxed);
+      return PoolStatus{old};
+    }
+
+    /// Returns *assumed* new status, including the just performed +1.
+    PoolStatus statusMarkWaitingAssumeAcquire() {
+      auto old = status.fetch_or(PoolStatus::waiting,
+                                 std::memory_order_acquire);
+      return PoolStatus{old | PoolStatus::waiting};
+    }
+
+    PoolStatus statusRemoveWaitingRelease() {
+      auto old = status.fetch_and(~PoolStatus::waiting,
+                                  std::memory_order_release);
+      return PoolStatus{old};
+    }
+
+    /// Decrement the pending task count.
+    /// Returns *assumed* new status, including the just performed +1.
+    PoolStatus statusDecrementPendingAssumeAcquire() {
+      auto old = status.fetch_sub(PoolStatus::onePendingTask,
+                                  std::memory_order_acquire);
+      assert(PoolStatus{old}.pendingTasks() > 0 && "attempted to decrement pending count when it was 0 already");
+      return PoolStatus{old - PoolStatus::onePendingTask};
+    }
+
+    /// Increment the pending task count.
+    ///
+    /// Returns *assumed* new status, including the just performed -1.
+    PoolStatus statusIncrementPendingAssumeAcquire() {
+      auto old = status.fetch_add(PoolStatus::onePendingTask,
+                                  std::memory_order_acquire);
+      return PoolStatus{old + PoolStatus::onePendingTask};
+    }
+
+    /// Similar to decrementing the pending count, however does so 'relaxed'.
+    /// Used to undo an optimistic increment, when the pool already is cancelled.
+    ///
+    /// Returns *assumed* new status, including the just performed -1.
+    PoolStatus statusUndoIncrementPendingAssumeRelaxed() {
+      auto o = status.fetch_sub(PoolStatus::onePendingTask,
+                                std::memory_order_relaxed);
+      return PoolStatus{o - PoolStatus::onePendingTask};
+    }
+
+    /// Add a single pending task to the status counter.
+    /// This is used to implement next() properly, as we need to know if there
+    /// are pending tasks worth suspending/waiting for or not.
+    ///
+    /// Note that the group does *not* store child tasks at all, as they are
+    /// stored in the `TaskPoolTaskStatusRecord` inside the current task, that
+    /// is currently executing the group. Here we only need the counts of
+    /// pending/ready tasks.
+    ///
+    /// If the `unconditionally` parameter is `true` the operation always successfully
+    /// adds a pending task, even if the group is cancelled. If the unconditionally
+    /// flag is `false`, the added pending count will be *reverted* before returning.
+    /// This is because we will NOT add a task to a cancelled group, unless doing
+    /// so unconditionally.
+    ///
+    /// Returns *assumed* new status, including the just performed +1.
```

Comment:
- [ ] TODO: go over all comments again and remove mentions of group and make sure the comments actually cover the details of the `Pool` impl